File: ios/GuideStreamTV/Services/RokuECPClient.swift
//
//  RokuECPClient.swift
//  GuideStreamTV
//
//  Sends launch commands to a Roku device using the External Control Protocol.
//  Docs: https://developer.roku.com/docs/developer-program/dev-tools/external-control-api.md
//
//  IMPORTANT: ECP is plain HTTP on port 8060. iOS App Transport Security blocks
//  cleartext requests to LAN IP literals (192.168.x.x) through URLSession by
//  default — those requests die before they leave the device with no useful
//  error. We bypass ATS by writing the HTTP/1.0 POST by hand over a raw
//  NWConnection TCP socket. NWConnection isn't governed by ATS, so cleartext
//  LAN traffic goes through cleanly.
//
//  Roku OS 14.1+ note: As of December 2024, the "Network access" sub-setting
//  under Settings → System → Advanced → Control by mobile apps defaults to
//  "Limited mode", which causes the Roku to reject `launch` and `keypress`
//  requests with HTTP 403 from devices it hasn't paired with via the Roku
//  app. This is by far the most common reason cast-to-Roku fails today, so
//  the launch result type explicitly distinguishes that case so the UI can
//  guide users to flip the setting to "Permissive" or "Enabled".
//

import Foundation
import Network

enum RokuChannel {
    /// Maps a streaming platform label to the Roku channel ID.
    /// Normalises dashes, underscores, and parentheses before matching
    /// so strings like "netflix-4k", "Amazon Prime Video", or "prime_video"
    /// all resolve correctly.
    static func id(for platform: String) -> String? {
        let key = platform
            .lowercased()
            .replacingOccurrences(of: "_", with: " ")
            .replacingOccurrences(of: "-", with: " ")
            .components(separatedBy: CharacterSet.alphanumerics.union(.whitespaces).inverted)
            .joined()

        if key.contains("netflix")              { return "12" }
        if key.contains("hbo") || key.contains(" max") || key.hasSuffix("max") { return "61322" }
        if key.contains("hulu")                { return "2285" }
        if key.contains("disney")              { return "291097" }
        if key.contains("prime") || key.contains("amazon") { return "13" }
        if key.contains("apple tv") || key.contains("appletv") || key.contains("apple ") { return "551012" }
        if key.contains("paramount")           { return "31440" }
        if key.contains("peacock")             { return "593099" }
        if key.contains("youtube tv") || key.contains("youtubetv") { return "195316" }
        if key.contains("youtube")             { return "837" }
        if key.contains("showtime")            { return "60308" }
        if key.contains("starz")               { return "151908" }
        if key.contains("crunchyroll")         { return "55307" }
        if key.contains("tubi")                { return "41468" }
        if key.contains("pluto")               { return "74519" }
        if key.contains("plex")                { return "13535" }
        return nil
    }
}

/// Outcome of a Roku ECP launch attempt. Lets the UI tell the user *why*
/// a launch failed instead of just "couldn't reach device".
enum RokuLaunchResult: Equatable {
    /// The channel opened (Roku replied 2xx).
    case ok
    /// Roku replied 401/403 — almost always means the user has
    /// "Network access" set to Limited mode under
    /// Settings → System → Advanced system settings → Control by mobile apps.
    /// The fix is to switch it to "Permissive" or "Enabled".
    case limitedMode
    /// Roku is reachable but rejected the launch (404 unknown channel id,
    /// 500, etc.). The associated value is the HTTP status when available.
    case rejected(Int)
    /// The HTTP request never completed — wrong host, device offline,
    /// AP isolation, VPN blocking LAN traffic, etc.
    case unreachable

    var isSuccess: Bool { self == .ok }
}

/// Lower-level result for a single raw HTTP request — exposes the HTTP
/// status code so `launch` can classify the failure.
private enum RokuHTTPOutcome {
    case status(Int)
    case timeout
    case socketFailure
}

enum RokuECPClient {

    // MARK: - ECP status check (pre-launch)

    /// Outcome of probing `/query/device-info` before attempting a launch.
    /// Lets the UI route to the limited-mode help screen *before* we burn
    /// the Home-keypress + 1.5s sleep on a device that will always 403.
    enum ECPStatus { case enabled, limited, unreachable }

    /// Checks whether ECP is enabled on the Roku device by probing
    /// /query/device-info via raw HTTP GET.
    ///
    /// - `.enabled`  — ECP is working, proceed with launch.
    /// - `.limited`  — Device is in limited mode (403 response). User must
    ///   enable "Control by mobile apps" in Roku settings.
    /// - `.unreachable` — No response within timeout (wrong IP, device off,
    ///   firewall, or different network).
    nonisolated static func checkECPEnabled(host: String, port: UInt16) async -> ECPStatus {
        await withCheckedContinuation { (continuation: CheckedContinuation<ECPStatus, Never>) in
            let nwHost = NWEndpoint.Host(host)
            guard let nwPort = NWEndpoint.Port(rawValue: port) else {
                continuation.resume(returning: .unreachable); return
            }
            let params = NWParameters.tcp
            params.prohibitedInterfaceTypes = [.cellular]
            let conn = NWConnection(host: nwHost, port: nwPort, using: params)

            let lock = NSLock()
            var didResume = false
            let finish: (ECPStatus) -> Void = { value in
                lock.lock()
                let already = didResume
                didResume = true
                lock.unlock()
                if already { return }
                conn.cancel()
                continuation.resume(returning: value)
            }

            DispatchQueue.global().asyncAfter(deadline: .now() + 3.0) {
                finish(.unreachable)
            }

            conn.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    let req = "GET /query/device-info HTTP/1.0\r\n"
                        + "Host: \(host)\r\n"
                        + "Connection: close\r\n\r\n"
                    conn.send(content: req.data(using: .utf8), completion: .contentProcessed { _ in })
                    var buffer = Data()
                    func readNext() {
                        conn.receive(minimumIncompleteLength: 1, maximumLength: 4096) { data, _, isComplete, error in
                            if let data = data { buffer.append(data) }
                            if let line = firstResponseLine(in: buffer) {
                                if line.contains(" 403") {
                                    finish(.limited)
                                } else if line.contains(" 2") {
                                    finish(.enabled)
                                } else {
                                    finish(.unreachable)
                                }
                                return
                            }
                            if isComplete || error != nil { finish(.unreachable); return }
                            readNext()
                        }
                    }
                    readNext()
                case .failed, .cancelled:
                    finish(.unreachable)
                default:
                    break
                }
            }
            conn.start(queue: .global(qos: .userInitiated))
        }
    }

    /// Fires an ECP launch request at the Roku device.
    ///
    /// `contentId` is the channel-specific catalog identifier — we pass the
    /// TMDB id as a best-effort hint; most first-party channels (Netflix,
    /// Hulu, Max, etc.) use their own internal IDs and will ignore an
    /// unrecognised value, simply opening to their landing screen. For
    /// channels that DO accept arbitrary identifiers (Jellyfin, Plex,
    /// sideloaded apps) the deeplink lands directly on the title.
    ///
    /// Attempt order:
    ///   1. `/launch/<id>?contentId=<id>&mediaType=<type>` — modern lowercase
    ///      params per Roku's current developer docs.
    ///   2. `/launch/<id>?contentID=<id>&MediaType=<type>` — legacy capitalised
    ///      params; some older channels expected this exact case.
    ///   3. `/launch/<id>` — bare launch, opens the channel home. This is the
    ///      "did we reach the box at all?" probe.
    ///
    /// Sends `/keypress/Home` first so the channel always launches from a
    /// known state, then waits for the Home transition to finish before
    /// sending the launch command. On TCL Roku TVs the launch is silently
    /// dropped if it arrives mid-animation. On Roku OS 14.1+ in Limited mode
    /// the Home keypress returns 403, but that's harmless — the launch still
    /// proceeds and either succeeds (channel opens over whatever was on
    /// screen) or returns `.limitedMode`.
    /// Launches a Roku channel using a Watchmode `roku_url` path — a relative
    /// ECP launch string of the form `launch/{channelId}?contentID={contentId}`.
    /// Watchmode resolves these per-source on its paid plan; on the free plan
    /// the field is `nil` and callers should fall back to `launch(host:port:channelId:contentId:mediaType:)`.
    ///
    /// Normalisation handles both bare relative paths and full URLs that happen
    /// to contain a `launch/` segment (e.g. `https://therokuchannel.roku.com/...launch/2285?...`).
    /// If the input can't be reduced to a usable ECP path, the method returns
    /// `.unreachable` without making any request so the caller can fall back.
    nonisolated static func launch(
        host: String,
        port: UInt16,
        rokuURLPath rawPath: String
    ) async -> RokuLaunchResult {
        // Normalise the input — strip schemes, keep only the ECP path portion.
        var normalized = rawPath.trimmingCharacters(in: .whitespacesAndNewlines)
        if normalized.contains("://"),
           let launchRange = normalized.range(of: "launch/") {
            normalized = String(normalized[launchRange.lowerBound...])
        }
        if normalized.hasPrefix("/") {
            normalized = String(normalized.dropFirst())
        }
        guard normalized.contains("launch/") else {
            return .unreachable
        }

        // Send Home first so the channel always launches from a known state.
        _ = await rawHTTPPost(host: host, port: port, path: "/keypress/Home", timeout: 2.0)
        try? await Task.sleep(for: .milliseconds(1200))

        let outcome = await rawHTTPPost(host: host, port: port, path: "/" + normalized, timeout: 4.0)
        switch outcome {
        case .status(let code):
            if (200..<300).contains(code) {
                return .ok
            }
            if code == 401 || code == 403 {
                return .limitedMode
            }
            // Bare-launch fallback: parse the channel ID between "launch/" and
            // the first "?" (or end of string) and open the channel home.
            let channelPart = normalized
                .replacingOccurrences(of: "launch/", with: "")
            let channelId = channelPart.components(separatedBy: "?").first ?? channelPart
            let fallbackOutcome = await rawHTTPPost(
                host: host, port: port,
                path: "/launch/\(channelId)", timeout: 4.0
            )
            if case .status(let fbCode) = fallbackOutcome, (200..<300).contains(fbCode) {
                return .ok
            }
            if case .status(let fbCode) = fallbackOutcome {
                return .rejected(fbCode)
            }
            return .rejected(code)
        case .timeout, .socketFailure:
            return .unreachable
        }
    }

    nonisolated static func launch(
        host: String,
        port: UInt16,
        channelId: String,
        contentId: String? = nil,
        mediaType: String = "series"
    ) async -> RokuLaunchResult {
        // Send Home first so we always launch from a known state.
        // TCL Roku TVs silently drop the launch if it arrives while
        // the Home transition is still animating. That animation takes
        // 1.2–1.8s, so we wait 1500ms as the safe minimum — at 900ms the
        // launch could still land mid-transition and the contentId gets
        // dropped, opening the channel home screen instead of the title.
        _ = await rawHTTPPost(host: host, port: port, path: "/keypress/Home", timeout: 2.0)
        try? await Task.sleep(for: .milliseconds(1500))

        // Build the candidate paths in priority order.
        var paths: [String] = []
        if let contentId, !contentId.isEmpty {
            let cid = contentId.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? contentId
            let mt = mediaType.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? mediaType
            paths.append("/launch/\(channelId)?contentId=\(cid)&mediaType=\(mt)")
            paths.append("/launch/\(channelId)?contentID=\(cid)&MediaType=\(mt)")
        }
        paths.append("/launch/\(channelId)")

        var sawForbidden = false
        var lastNonOkStatus: Int? = nil

        for path in paths {
            let outcome = await rawHTTPPost(host: host, port: port, path: path, timeout: 4.0)
            switch outcome {
            case .status(let code):
                if (200..<300).contains(code) {
                    return .ok
                }
                if code == 401 || code == 403 {
                    sawForbidden = true
                } else {
                    lastNonOkStatus = code
                }
            case .timeout, .socketFailure:
                // Network-level failure — try the next variant, but if every
                // attempt is a socket failure we'll fall through to .unreachable.
                continue
            }
        }

        if sawForbidden { return .limitedMode }
        if let lastNonOkStatus { return .rejected(lastNonOkStatus) }
        return .unreachable
    }

    /// Sends `keypress/<key>` to the Roku — useful for sending Home/Back
    /// after a launch (e.g. dismissing the active session via the home
    /// banner). Best-effort; returns `false` for any non-2xx response or
    /// network failure. Public so `CastPlaybackState` can reuse it.
    ///
    /// NOTE: On Roku OS 14.1+, this returns 403 unless the user has set
    /// "Network access" to "Permissive" or "Enabled". Callers should treat
    /// failure here as advisory, not fatal.
    nonisolated static func keypress(host: String, port: UInt16, key: String) async -> Bool {
        let outcome = await rawHTTPPost(host: host, port: port, path: "/keypress/\(key)", timeout: 2.5)
        if case .status(let code) = outcome, (200..<300).contains(code) {
            return true
        }
        return false
    }

    /// Cheap reachability probe — sends `GET /query/device-info` (always
    /// allowed by Roku, even in Limited mode) and returns true if the device
    /// replies with anything. Used to differentiate "device offline / wrong
    /// IP" from "device online but rejecting commands".
    nonisolated static func isReachable(host: String, port: UInt16) async -> Bool {
        await withCheckedContinuation { (continuation: CheckedContinuation<Bool, Never>) in
            let nwHost = NWEndpoint.Host(host)
            guard let nwPort = NWEndpoint.Port(rawValue: port) else {
                continuation.resume(returning: false); return
            }
            let params = NWParameters.tcp
            params.prohibitedInterfaceTypes = [.cellular]
            let conn = NWConnection(host: nwHost, port: nwPort, using: params)

            let lock = NSLock()
            var didResume = false
            let finish: (Bool) -> Void = { value in
                lock.lock()
                let already = didResume
                didResume = true
                lock.unlock()
                if already { return }
                conn.cancel()
                continuation.resume(returning: value)
            }
            DispatchQueue.global().asyncAfter(deadline: .now() + 2.0) { finish(false) }

            conn.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    let req = "GET /query/device-info HTTP/1.0\r\nHost: \(host)\r\nConnection: close\r\n\r\n"
                    conn.send(content: req.data(using: .utf8), completion: .contentProcessed { _ in })
                    conn.receive(minimumIncompleteLength: 1, maximumLength: 256) { data, _, _, _ in
                        finish(!(data?.isEmpty ?? true))
                    }
                case .failed, .cancelled:
                    finish(false)
                default:
                    break
                }
            }
            conn.start(queue: .global(qos: .userInitiated))
        }
    }

    /// Extracts the platform-native content ID from a Watchmode web_url.
    /// This is the ID Roku ECP needs for deep linking into a specific title.
    ///
    /// Examples:
    /// https://www.netflix.com/title/80057281 → "80057281"
    /// https://www.amazon.com/dp/B01N4AH0XV → "B01N4AH0XV"
    /// https://www.hulu.com/series/ancient-aliens-xx → "ancient-aliens-xx"
    /// https://www.disneyplus.com/series/foo/barId → "barId"
    /// https://www.max.com/shows/foo/uuid-here → "uuid-here"
    nonisolated static func extractContentId(fromWebURL urlString: String, platform: String) -> String? {
        guard let comps = URLComponents(string: urlString),
              let host = comps.host else { return nil }
        let path = comps.path
        let hostLower = host.lowercased()
        let key = platform.lowercased()

        // Netflix: /title/{id}
        if hostLower.contains("netflix") || key.contains("netflix") {
            let parts = path.split(separator: "/").map(String.init)
            if let idx = parts.firstIndex(of: "title"), idx + 1 < parts.count {
                return parts[idx + 1]
            }
            return parts.last
        }

        // Amazon Prime Video: /dp/{ASIN} or /gp/video/detail/{ASIN}
        if hostLower.contains("amazon") || key.contains("prime") || key.contains("amazon") {
            let parts = path.split(separator: "/").map(String.init)
            if let idx = parts.firstIndex(of: "dp"), idx + 1 < parts.count {
                return parts[idx + 1]
            }
            if let idx = parts.firstIndex(of: "detail"), idx + 1 < parts.count {
                return parts[idx + 1]
            }
            return parts.last
        }

        // Hulu: /series/{slug} or /movie/{slug} or /watch/{id}
        if hostLower.contains("hulu") || key.contains("hulu") {
            let parts = path.split(separator: "/").map(String.init)
            if let idx = parts.firstIndex(where: { $0 == "series" || $0 == "movie" || $0 == "watch" }),
               idx + 1 < parts.count {
                return parts[idx + 1]
            }
            return parts.last
        }

        // Disney+: last non-empty path component is the content ID
        if hostLower.contains("disneyplus") || key.contains("disney") {
            return path.split(separator: "/").last.map(String.init)
        }

        // Max (HBO Max): last path component is UUID
        if hostLower.contains("max.com") || key.contains("hbo") || key.contains(" max") {
            return path.split(separator: "/").last.map(String.init)
        }

        // Paramount+: /shows/{slug} or /movies/{slug}
        if hostLower.contains("paramount") || key.contains("paramount") {
            let parts = path.split(separator: "/").map(String.init)
            if let idx = parts.firstIndex(where: { $0 == "shows" || $0 == "movies" }),
               idx + 1 < parts.count {
                return parts[idx + 1]
            }
            return parts.last
        }

        // Peacock, Apple TV+, and others: last path component
        return path.split(separator: "/").last.map(String.init)
    }

    // MARK: - Raw HTTP POST (ATS-bypass)

    /// Performs an HTTP/1.0 POST with an empty body via raw NWConnection.
    /// Returns the parsed HTTP status code, or a non-status outcome when the
    /// connection itself failed (timeout, refused, etc.).
    ///
    /// Implementation notes:
    ///   - Roku ECP requires POST (with `-d ''` in curl docs). The `Content-Length: 0`
    ///     header is mandatory; without it Roku waits for body bytes and times out.
    ///   - We use HTTP/1.0 + `Connection: close` so the device terminates the
    ///     stream as soon as it's done writing — we don't need keep-alive.
    nonisolated private static func rawHTTPPost(host: String, port: UInt16, path: String, timeout: TimeInterval) async -> RokuHTTPOutcome {
        await withCheckedContinuation { (continuation: CheckedContinuation<RokuHTTPOutcome, Never>) in
            let nwHost = NWEndpoint.Host(host)
            guard let nwPort = NWEndpoint.Port(rawValue: port) else {
                continuation.resume(returning: .socketFailure); return
            }
            let params = NWParameters.tcp
            params.prohibitedInterfaceTypes = [.cellular]
            let conn = NWConnection(host: nwHost, port: nwPort, using: params)

            let lock = NSLock()
            var didResume = false
            let finish: (RokuHTTPOutcome) -> Void = { value in
                lock.lock()
                let already = didResume
                didResume = true
                lock.unlock()
                if already { return }
                conn.cancel()
                continuation.resume(returning: value)
            }

            DispatchQueue.global().asyncAfter(deadline: .now() + timeout) {
                finish(.timeout)
            }

            conn.stateUpdateHandler = { state in
                switch state {
                case .ready:
                    let req = "POST \(path) HTTP/1.0\r\n"
                        + "Host: \(host)\r\n"
                        + "User-Agent: GuideStreamTV-ECP/1.0\r\n"
                        + "Content-Length: 0\r\n"
                        + "Connection: close\r\n\r\n"
                    conn.send(content: req.data(using: .utf8), completion: .contentProcessed { error in
                        if error != nil { finish(.socketFailure); return }
                    })
                    var buffer = Data()
                    func readNext() {
                        conn.receive(minimumIncompleteLength: 1, maximumLength: 4096) { data, _, isComplete, error in
                            if let data = data, !data.isEmpty { buffer.append(data) }
                            // Once we have any status line, we can decide.
                            if let code = parseStatusCode(in: buffer) {
                                finish(.status(code))
                                return
                            }
                            if isComplete || error != nil || buffer.count > 16 * 1024 {
                                finish(.socketFailure)
                                return
                            }
                            readNext()
                        }
                    }
                    readNext()
                case .failed, .cancelled:
                    finish(.socketFailure)
                default:
                    break
                }
            }
            conn.start(queue: .global(qos: .userInitiated))
        }
    }

    /// Parses the HTTP status code from a buffer that may contain a partial
    /// or complete response. Returns nil if the status line hasn't fully
    /// arrived yet (no CRLF terminator). Expects an HTTP/1.x response line
    /// like `HTTP/1.0 200 OK` or `HTTP/1.1 403 Forbidden`.
    nonisolated private static func parseStatusCode(in data: Data) -> Int? {
        guard let line = firstResponseLine(in: data) else { return nil }
        // Format: HTTP/<ver> <code> <reason>
        let parts = line.split(separator: " ", maxSplits: 2, omittingEmptySubsequences: true)
        guard parts.count >= 2, let code = Int(parts[1]) else { return nil }
        return code
    }

    /// Returns the first line of an HTTP response once enough bytes have
    /// arrived for a CRLF to appear, or `nil` if the line isn't complete yet.
    nonisolated private static func firstResponseLine(in data: Data) -> String? {
        let crlf: [UInt8] = [0x0d, 0x0a]
        guard data.count >= 2 else { return nil }
        var i = 0
        let upper = data.count - 2
        while i <= upper {
            if data[i] == crlf[0] && data[i + 1] == crlf[1] {
                return String(data: data.subdata(in: 0..<i), encoding: .utf8)
            }
            i += 1
        }
        return nil
    }
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/RokuEcpClient.kt
package com.rork.guidestreamtvandroid.data.remote

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import kotlinx.coroutines.withTimeoutOrNull
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.InetSocketAddress
import java.net.Socket
import java.net.URL
import java.net.URLEncoder

/**
 * Sends launch commands to a Roku device using the External Control Protocol.
 * Mirrors iOS `RokuECPClient.swift` one-for-one.
 *
 * ECP is plain HTTP on port 8060. Raw `java.net.Socket` is used because
 * Android's cleartext-traffic policy would block OkHttp or HttpURLConnection
 * calls to LAN IP literals.
 */
object RokuEcpClient {

    /** Maps a streaming platform label to the Roku channel ID. */
    object RokuChannel {
        fun id(platform: String): String? {
            val key = platform.lowercase()
                .replace("_", " ")
                .replace("-", " ")
                .filter { it.isLetterOrDigit() || it.isWhitespace() }

            if (key.contains("netflix")) return "12"
            if (key.contains("hbo") || key.contains(" max") || key.endsWith("max")) return "61322"
            if (key.contains("hulu")) return "2285"
            if (key.contains("disney")) return "291097"
            if (key.contains("prime") || key.contains("amazon")) return "13"
            if (key.contains("apple tv") || key.contains("appletv") || key.contains("apple ")) return "551012"
            if (key.contains("paramount")) return "31440"
            if (key.contains("peacock")) return "593099"
            if (key.contains("youtube tv") || key.contains("youtubetv")) return "195316"
            if (key.contains("youtube")) return "837"
            if (key.contains("showtime")) return "60308"
            if (key.contains("starz")) return "151908"
            if (key.contains("crunchyroll")) return "55307"
            if (key.contains("tubi")) return "41468"
            if (key.contains("pluto")) return "74519"
            if (key.contains("plex")) return "13535"
            return null
        }
    }

    /** Outcome of probing `/query/device-info` before attempting a launch. */
    enum class EcpStatus { ENABLED, LIMITED, UNREACHABLE }

    /** Outcome of a Roku ECP launch attempt. */
    sealed interface RokuLaunchResult {
        data object Ok : RokuLaunchResult
        data object LimitedMode : RokuLaunchResult
        data class Rejected(val status: Int) : RokuLaunchResult
        data object Unreachable : RokuLaunchResult
    }

    /**
     * Checks whether ECP is enabled on the Roku device. A device that answers
     * with any HTTP status is treated as `.ENABLED` — the launch path will
     * classify the real failure. Only a timeout or socket failure returns
     * `.UNREACHABLE`.
     */
    suspend fun checkEcpEnabled(host: String, port: Int): EcpStatus {
        val response = rawHttpRequest(host, port, "GET", "/query/device-info", 3000L)
        if (response == null) return EcpStatus.UNREACHABLE
        val code = parseStatusCode(response)
        return when {
            code == null -> EcpStatus.UNREACHABLE
            code in 200..299 -> EcpStatus.ENABLED
            code == 401 || code == 403 -> EcpStatus.LIMITED
            else -> EcpStatus.ENABLED
        }
    }

    /**
     * Launches a Roku channel using a Watchmode `roku_url` path.
     * Sends `/keypress/Home` first, then tries the launch path, falling back
     * to a bare `/launch/{channelId}` if the parameterised launch is rejected.
     */
    suspend fun launch(host: String, port: Int, rokuUrlPath: String): RokuLaunchResult {
        var normalized = rokuUrlPath.trim()
        if (normalized.contains("://")) {
            val launchIndex = normalized.indexOf("launch/")
            if (launchIndex >= 0) {
                normalized = normalized.substring(launchIndex)
            }
        }
        if (normalized.startsWith("/")) {
            normalized = normalized.drop(1)
        }
        if (!normalized.contains("launch/")) {
            return RokuLaunchResult.Unreachable
        }

        rawHttpPost(host, port, "/keypress/Home", 1200L)
        delay(1200)

        val outcome = rawHttpPost(host, port, "/$normalized", 4000L)
        if (outcome != null) {
            if (outcome in 200..299) return RokuLaunchResult.Ok
            if (outcome == 401 || outcome == 403) return RokuLaunchResult.LimitedMode

            val channelPart = normalized.replace("launch/", "")
            val channelId = channelPart.split("?").firstOrNull() ?: channelPart
            val fallbackOutcome = rawHttpPost(host, port, "/launch/$channelId", 4000L)
            if (fallbackOutcome != null) {
                if (fallbackOutcome in 200..299) return RokuLaunchResult.Ok
                return RokuLaunchResult.Rejected(fallbackOutcome)
            }
            return RokuLaunchResult.Rejected(outcome)
        }
        return RokuLaunchResult.Unreachable
    }

    /**
     * Launches a Roku channel by channel ID with an optional content ID.
     * Tries parameterised launch paths first, then falls back to bare launch.
     */
    suspend fun launch(
        host: String,
        port: Int,
        channelId: String,
        contentId: String?,
        mediaType: String,
    ): RokuLaunchResult {
        rawHttpPost(host, port, "/keypress/Home", 1200L)
        delay(1500)

        val paths = mutableListOf<String>()
        if (contentId != null && contentId.isNotEmpty()) {
            val cid = URLEncoder.encode(contentId, "UTF-8")
            val mt = URLEncoder.encode(mediaType, "UTF-8")
            paths.add("/launch/$channelId?contentId=$cid&mediaType=$mt")
            paths.add("/launch/$channelId?contentID=$cid&MediaType=$mt")
        }
        paths.add("/launch/$channelId")

        var sawForbidden = false
        var lastNonOkStatus: Int? = null

        for (path in paths) {
            val outcome = rawHttpPost(host, port, path, 4000L)
            if (outcome != null) {
                if (outcome in 200..299) return RokuLaunchResult.Ok
                if (outcome == 401 || outcome == 403) {
                    sawForbidden = true
                } else {
                    lastNonOkStatus = outcome
                }
            }
        }

        if (sawForbidden) return RokuLaunchResult.LimitedMode
        if (lastNonOkStatus != null) return RokuLaunchResult.Rejected(lastNonOkStatus)
        return RokuLaunchResult.Unreachable
    }

    /** Sends `keypress/<key>` to the Roku. Best-effort; returns false for non-2xx. */
    suspend fun keypress(host: String, port: Int, key: String): Boolean {
        val code = rawHttpPost(host, port, "/keypress/$key", 2500L)
        return code != null && code in 200..299
    }

    /** Cheap reachability probe — GET /query/device-info, returns true if any response. */
    suspend fun isReachable(host: String, port: Int): Boolean {
        return rawHttpRequest(host, port, "GET", "/query/device-info", 2000L) != null
    }

    /**
     * Extracts the platform-native content ID from a Watchmode web_url.
     * Reproduces the Netflix, Amazon, Hulu, Disney+, Max, Paramount+ and
     * default path-parsing rules from the iOS RokuECPClient.
     */
    fun extractContentId(fromWebUrl: String, platform: String): String? {
        val url = try { URL(fromWebUrl) } catch (_: Exception) { return null }
        val host = url.host?.lowercase() ?: return null
        val path = url.path ?: ""
        val key = platform.lowercase()

        val parts = path.split("/").filter { it.isNotEmpty() }

        if (host.contains("netflix") || key.contains("netflix")) {
            val idx = parts.indexOf("title")
            if (idx >= 0 && idx + 1 < parts.size) return parts[idx + 1]
            return parts.lastOrNull()
        }

        if (host.contains("amazon") || key.contains("prime") || key.contains("amazon")) {
            val idx = parts.indexOf("dp")
            if (idx >= 0 && idx + 1 < parts.size) return parts[idx + 1]
            val detailIdx = parts.indexOf("detail")
            if (detailIdx >= 0 && detailIdx + 1 < parts.size) return parts[detailIdx + 1]
            return parts.lastOrNull()
        }

        if (host.contains("hulu") || key.contains("hulu")) {
            val idx = parts.indexOfFirst { it == "series" || it == "movie" || it == "watch" }
            if (idx >= 0 && idx + 1 < parts.size) return parts[idx + 1]
            return parts.lastOrNull()
        }

        if (host.contains("disneyplus") || key.contains("disney")) {
            return parts.lastOrNull()
        }

        if (host.contains("max.com") || key.contains("hbo") || key.contains(" max")) {
            return parts.lastOrNull()
        }

        if (host.contains("paramount") || key.contains("paramount")) {
            val idx = parts.indexOfFirst { it == "shows" || it == "movies" }
            if (idx >= 0 && idx + 1 < parts.size) return parts[idx + 1]
            return parts.lastOrNull()
        }

        return parts.lastOrNull()
    }

    // MARK: - Raw HTTP (cleartext bypass)

    private suspend fun rawHttpRequest(
        host: String,
        port: Int,
        method: String,
        path: String,
        timeoutMs: Long,
    ): String? = withContext(Dispatchers.IO) {
        withTimeoutOrNull(timeoutMs) {
            try {
                val socket = Socket()
                socket.connect(InetSocketAddress(host, port), timeoutMs.toInt())
                socket.soTimeout = timeoutMs.toInt()

                val request = buildString {
                    append("$method $path HTTP/1.0\r\n")
                    append("Host: $host\r\n")
                    append("User-Agent: GuideStreamTV-ECP/1.0\r\n")
                    if (method == "POST") {
                        append("Content-Length: 0\r\n")
                    }
                    append("Connection: close\r\n\r\n")
                }
                socket.getOutputStream().write(request.toByteArray())
                socket.getOutputStream().flush()

                val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
                val response = StringBuilder()
                var line: String?
                while (reader.readLine().also { line = it } != null) {
                    response.append(line).append("\n")
                    if (response.length > 16384) break
                }
                socket.close()
                response.toString()
            } catch (_: Exception) {
                null
            }
        }
    }

    private suspend fun rawHttpPost(
        host: String,
        port: Int,
        path: String,
        timeoutMs: Long,
    ): Int? {
        val response = rawHttpRequest(host, port, "POST", path, timeoutMs) ?: return null
        return parseStatusCode(response)
    }

    private fun parseStatusCode(response: String): Int? {
        val firstLine = response.lines().firstOrNull() ?: return null
        val parts = firstLine.trim().split(" ")
        if (parts.size >= 2) {
            return parts[1].toIntOrNull()
        }
        return null
    }
}
